Description: Fix wxSocket::WaitForAccept() in worker thread.
 This was broken because Select() never returned wxSOCKET_CONNECTION_FLAG which
 is supposed to be set when a connection is accepted.
Origin: upstream, http://trac.wxwidgets.org/changeset/75211
Bug: http://trac.wxwidgets.org/ticket/15669
Applied-Upstream: http://trac.wxwidgets.org/changeset/75211
Last-Update: 2013-11-30

Index: /wxWidgets/branches/WX_3_0_BRANCH/src/common/socket.cpp
===================================================================
--- a/src/common/socket.cpp (revision 74602)
+++ b/src/common/socket.cpp (revision 75211)
@@ -1358,5 +1358,23 @@
     wxSocketEventFlags detected = 0;
     if ( preadfds && wxFD_ISSET(m_fd, preadfds) )
-        detected |= wxSOCKET_INPUT_FLAG;
+    {
+        // check for the case of a server socket waiting for connection
+        if ( m_server && (flags & wxSOCKET_CONNECTION_FLAG) )
+        {
+            int error;
+            SOCKOPTLEN_T len = sizeof(error);
+            m_establishing = false;
+            getsockopt(m_fd, SOL_SOCKET, SO_ERROR, (char*)&error, &len);
+
+            if ( error )
+                detected = wxSOCKET_LOST_FLAG;
+            else
+                detected |= wxSOCKET_CONNECTION_FLAG;
+        }
+        else // not called to get non-blocking accept() status
+        {
+            detected |= wxSOCKET_INPUT_FLAG;
+        }
+    }
 
     if ( pwritefds && wxFD_ISSET(m_fd, pwritefds) )
